File: JavaScript/src/items/TimeSensitiveItem.js
import Item from './Item';

/* Represents items whose quality goes up increasingly rapidly until they are
 * deemed worthless */
export default class TimeSensitiveItem extends Item {
  tick() {
    this.gildedRose.daysRemaining -= 1;

    if (this.gildedRose.daysRemaining < 0) {
      this.gildedRose.quality = 0;
      return;
    }

    this.gildedRose.quality += 1;
    if (this.gildedRose.daysRemaining < 10) {
      this.gildedRose.quality += 1;
    }
    if (this.gildedRose.daysRemaining < 5) {
      this.gildedRose.quality += 1;
    }

    this.ensureBoundedQuality();
  }
}

File: Ruby/lib/items/time_sensitive_item.rb
# frozen_string_literal: true

require 'items/item'

module Items
  # Represents items whose quality goes up increasingly rapidly until they are
  # deemed worthless
  class TimeSensitiveItem < Item
    def tick
      @gilded_rose.days_remaining -= 1

      return @gilded_rose.quality = 0 if @gilded_rose.days_remaining < 0

      @gilded_rose.quality += 1
      @gilded_rose.quality += 1 if @gilded_rose.days_remaining < 10
      @gilded_rose.quality += 1 if @gilded_rose.days_remaining < 5

      ensure_bounded_quality
    end
  end
end
